# EBA3400 Exam

## Task 1)  Password valdiation

In this task, I followed the SOLID principles, especially the Single Responsibility Principle, rather than making a "God validation function." Splitting up the functionalities of the password validator into multiple functions makes testing each condition easier. A disadvantage of my approach is a much higher time complexity (due to the number of loops).

In [22]:
# Checks if the length of the password is 6 to 12 characters. 
def length_check(password):
    password_length = len(password)
    if not (6 <= password_length <= 12):
        return False
    return True

# Checks if at least one character is uppercase 
def uppercase_check(password):
    if not length_check(password):
        return False
    for char in password:
        if char.isupper():
            return True
    return False

# Checks if at least one character is a digit
def digit_check(password):
    if not uppercase_check(password):
        return False
    for char in password:
        if char.isdigit():
            return True
    return False

# Checks if there are any special characters
def special_character_check(password):
    if not digit_check(password):
        return False
    special_characters = list("!@#$%^&*()-_=+[]{}|;:'\",.<>?")
    for char in password:
        for special_char in special_characters:
            if special_char in char:
                return False
    return True

# Validates the password by checking the conditions
def password_validator(password):
    if not special_character_check(password):
        print("Invalid password")
        return
    print("Valid password")


password_validator(input("Enter password:"))

Enter password:Password123
Valid password


## Task 2)  Data dispersion measurement

Again, I decided to separate logic into its own functions. First, my output displayed the wrong answer when I ran the dispersion function. Therefore, after studying the question more, I concluded that each element in the summation must be represented in its absolute form. I hope this was the correct conclusion. 

In [23]:
# Returns the average of elements in the list
def mean_of_list(number_list):
    size = len(number_list)
    counter = 0
    for number in number_list:
        counter += number
    return counter / size

# Returns the summation part of the calculation
def sum_of_list(number_list):
    mean = mean_of_list(number_list)
    counter = 0
    for number in number_list:
        x = abs(number - mean)
        counter += x
    return counter

# Returns the dispersion level of the list, rounded to two decimals
def dispersion(number_list):
    size = len(number_list)
    summation = sum_of_list(number_list)
    return round(summation / size, 2)


list_1 = [4, 1, 4, 1, 3, 5]
list_2 = [43, 31, 41, 40, 40, 48, 29, 48, 23, 33]

dispersion_1 = dispersion(list_1)
dispersion_2 = dispersion(list_2)

print(f"(1) list_1 gives a dispersion of {dispersion_1}")
print(f"(2) list_2 gives a dispersion of {dispersion_2}")

(1) list_1 gives a dispersion of 1.33
(2) list_2 gives a dispersion of 6.88


## Task 3)  Airline customer data

I had to separate each question into its own code block in this task. This is due to issues with printing information requested from the dataframe variable. 
I also added another optional way of solving the first question regarding the average age of the customers. 

In [24]:
import pandas as pd


df = pd.read_csv("airline.csv")

#Optional: We could read the value from df["Age"].describe().round(1)
average_age = round(df["Age"].mean(), 1)
print(f"The average age of the customers are {average_age} years old.")

The average age of the customers are 45.9 years old.


In [25]:
df["Gender"].value_counts()

Gender
Male      5036
Female    4964
Name: count, dtype: int64

As we can read from the text provided by performing the function "value_counts()," we can see the gender distribution as 5036 male and 4964 female. 

## Task 4)  Trip planer 

### 4.1 Bus stop information

I added the bus stop information in a dictionary, with the stop ID as the key and coordinates as the value. 
I also made an extra function for printing the values in the dictionary. 

In [26]:
bus_stop_dictionary = {
    1: [2, 1],
    2: [6, 1],
    3: [5, 3],
    4: [3, 4],
    5: [3, 6],
    6: [5, 7],
    7: [7, 6]
}

# Prints each key-value pair in the dictionary on separate lines. 
def print_dictionary(dictionary):
    for key in dictionary:
        bus_stop = dictionary[key]
        print(f"Location for bus stop {key} is {bus_stop}")


print_dictionary(bus_stop_dictionary)

Location for bus stop 1 is [2, 1]
Location for bus stop 2 is [6, 1]
Location for bus stop 3 is [5, 3]
Location for bus stop 4 is [3, 4]
Location for bus stop 5 is [3, 6]
Location for bus stop 6 is [5, 7]
Location for bus stop 7 is [7, 6]


### 4.2 Distance calculation

In this task, I made two functions—one for finding the Euclidean distance, and another for printing it. I noticed the coordinates were the same as bus stops 1 and 3. Therefore, I fetched the coordinates from the dictionary created in task 4.1

Also, the question never asked to round the number, so I did not. 

In [28]:
# Finds the euclidean distance between two locations
def find_euclidean_distance(loc1, loc2):
    x1 = loc1[0]
    y1 = loc1[1]
    x2 = loc2[0]
    y2 = loc2[1]

    return (((x2 - x1)**2) + ((y2 - y1)**2)) **0.5

# Find the locations by using the stop_ID's as keys to the dictionary created in task 4.1.
# Prints the message
def euclidean_distance_printer(stop_id_1, stop_id_2):
    loc1 = bus_stop_dictionary[stop_id_1]
    loc2 = bus_stop_dictionary[stop_id_2]
    euclidean_distance = find_euclidean_distance(loc1, loc2)
    print(f"The euclidean distance between bus stop {stop_id_1} and {stop_id_2} is {euclidean_distance}")


euclidean_distance_printer(1, 3)

The euclidean distance between bus stop 1 and 3 is 3.605551275463989


### 4.3 Nearest stop

I had some extra fun in this task, making two alternative solutions. 

* Reviewing everything, I realized I could sort the dictionary created in the first function. 

#### Solution 1)

Here, I made some extra functions, which allow me to obtain multiple stops in case some have the same Euclidean distance. I also had to do extra work formatting the message since the bus stop IDs are retrieved as a list.

In [30]:
# Uses the input locations to calculate euclidean distance between the user and all bus stops. 
# Returns a dictionary using the stop_ID as key, and euclidean distance as value. 
def euclidean_distances_to_dictionary(input_loc):
    distances = {}
    for key in bus_stop_dictionary:
        stop_loc = bus_stop_dictionary[key]
        distance = find_euclidean_distance(input_loc, stop_loc)
        distances[key] = distance
    return distances

# This function takes the dictionary created in the previous function, and loops over it.
# It uses the minimum value from the dictionary as a counter, to check if the distance is equal or lower.
# Returns a list of all the locations closest to the user.
def find_nearest_stops(input_loc):
    distances = euclidean_distances_to_dictionary(input_loc)
    stops = []
    counter = min(distances.values())

    for distance_key in distances:
        distance = distances[distance_key]
        if distance <= counter:
            stops.append(distance_key)
    return stops

# Formats the message, adjusting it after how many stops are equally close. 
def message_formatter(input_loc):
    stops = find_nearest_stops(input_loc)
    message = "The nearest "
    stops_length = len(stops)
    if stops_length == 1:
        message += "stop is Stop " + str(stops[0]) + "."
    elif stops_length > 1:
        loop_counter = 0
        for stop in stops:
            loop_counter += 1
            if loop_counter == 1:
                message += "stops are Stop: " + str(stop)
            elif loop_counter == stops_length:
                message += " and " + str(stop) + "."
            else:
                message += ", " + str(stop)
    return message

# Ask the user to input the coordinates, and prints the closest stop(s). 
def print_nearest_stops():
    input_x = input("Enter your coordinate (x):")
    input_y = input("Enter your coordinate (y):")
    message = message_formatter([float(input_x), float(input_y)])
    print(message)


print_nearest_stops()

Enter your coordinate (x):1
Enter your coordinate (y):5
The nearest stops are Stop: 4 and 5.


#### Solution 2)

A quick solution representing only one of the nearest bus stops. 

In [31]:
# This function uses a counter with a dummy value. Loops over the bus_stop dictionary created in task 4.1,
# And update the nearest_stop variable, if the distance is lower than the counter variable. 
# Returns the nearest bus stop ID. 
def alternative_nearest_stop(input_loc):
    counter = 10000000
    nearest_stop = 0
    for key in bus_stop_dictionary:
        stop_loc = bus_stop_dictionary[key]
        distance = find_euclidean_distance(input_loc, stop_loc)
        if distance < counter:
            nearest_stop = key
            counter = distance
    return nearest_stop

# Ask the user to input the coordinates, and prints the closest stop(s). 
def alternative_print_nearest_stop():
    input_x = input("Enter your coordinate (x):")
    input_y = input("Enter your coordinate (y):")
    nearest_stop = alternative_nearest_stop([float(input_x), float(input_y)])
    print(f"The nearest stop is Stop {nearest_stop}.")


alternative_print_nearest_stop()

Enter your coordinate (x):1
Enter your coordinate (y):5
The nearest stop is Stop 4.


### 4.4.Estimated travel time

I created a dictionary using the ID of the bus stop it departed from and the time it takes to reach the next bus stop as value. 

Then, I made two functions, one for calculating the estimated time and the other for printing the travel time. 

In [ ]:
travel_time_dictionary = {
    1: find_euclidean_distance([2, 1], [6, 1]) * 2.5,
    2: find_euclidean_distance([6, 1], [5, 3]) * 2.5,
    3: find_euclidean_distance([5, 3], [3, 4]) * 4,
    4: find_euclidean_distance([3, 4], [3, 6]) * 4,
    5: find_euclidean_distance([3, 6], [5, 7]) * 4,
    6: find_euclidean_distance([5, 7], [7, 6]) * 2.5
}

# First this function creates a list of all the bus stops between the departure and destination.
# Then it loops over the dictionary created above, and again loops over the list created. 
# Checks if the route ID, is the same as bus stop ID, if so it adds the travel time fetched from the dictionary. 
def estimated_time(departure, destination):
    travel_time = 0
    bus_stops = list(range(departure, destination))
    for routes in travel_time_dictionary:
        for stops in bus_stops:
            if routes == stops:
                travel_time += travel_time_dictionary[routes]
    return travel_time

# Ask the user to input departure and destination, and prints the travel time (rounded to 1 decimal). 
def print_estimated_time():
    departure = int(input("Departure stop:"))
    destination = int(input("Destination stop:"))
    travel_time = round(estimated_time(departure, destination), 1)
    print(f"Estimated bus travel time: {travel_time} min")


print_estimated_time()